File: assets/src/models/oauth_token.ts
export class OAuthToken {
  public name: string = ''

  public access: string = ''

  public email?: string

  public refresh?: string

  public expires_at?: Date | number

  public meta?: any

  public static fromJSON(obj: any) {
    let item: OAuthToken = Object.assign(new OAuthToken(), obj)

    if (obj.expires_at) item.expires_at = new Date(obj.expires_at)
    if (!item.meta) item.meta = {} as any

    return item
  }
}

export const GOOGLE_CAL = 'google-cal'

File: lib/sequence/oauth_tokens/oauth_token.ex
defmodule Sequence.OAuthTokens.OAuthToken do
  use Ecto.Schema
  import Ecto.Changeset

  @type t :: %__MODULE__{}

  schema "oauth_tokens" do
    field :email, :string
    field :access, :string
    field :deleted_at, :naive_datetime
    field :expires_at, :naive_datetime
    field :meta, :map
    field :name, :string
    field :refresh, :string

    belongs_to :user, Sequence.Users.User

    timestamps()
  end

  @doc false
  def changeset(oauth_token, attrs) do
    oauth_token
    |> cast(attrs, [:user_id, :email, :access, :expires_at, :deleted_at, :name, :refresh, :meta])
    |> validate_required([:user_id, :name])
  end
end
